File: java/src/br/main/tcpip/tratarMensagens/TratarRequisicoesBin.java
package br.main.tcpip.tratarMensagens;

import java.awt.Point;

import br.main.controles.Audio;
import br.main.controles.LockUnLockScreen;
import br.main.controles.MouseControl;
import br.main.controles.Shutdown;
import br.main.tcpip.interfaces.ITratarRequisicaoBin;
import br.main.util.BinaryUtil;
import br.main.util.MyBitSet;

public class TratarRequisicoesBin implements ITratarRequisicaoBin {

	private MouseControl mc = new MouseControl();
	
	@Override
	public MyBitSet tratar(MyBitSet mensagem) {
		if (mensagem == null || mensagem.size() <= 0) { return msgErro(); }
		
		MyBitSet cod = mensagem.slice(0, 3);
		System.out.println("[tratar] cod: " + cod.toString() + " = " + BinaryUtil.toInt(cod, true));
		
		switch(BinaryUtil.toInt(cod, true)) {
			case 0: return sinchronizar();
			case 1: return alterarVolume(mensagem);
			case 2: return desligarPC();
			case 3: return mouseMove(mensagem);
			case 4: return clickMouse();
			case 5: return lockScreen();
			case 6: return unLockScreen();
		}
		
		return msgErro();
	}

	private MyBitSet msgErro() {
		MyBitSet retorno = BinaryUtil.toMBitByte((byte)0x0, 1, false);
		retorno.append(BinaryUtil.toMBit("Erro"));
		return retorno;
	}
	
	private MyBitSet sinchronizar() {
		MyBitSet retorno = BinaryUtil.toMBitByte((byte)1, 1, false);
		try {
			float volume = Audio.getMasterOutputVolume() * 100f; // Linux - Audio
			System.out.println(String.format("Audio.getMasterOutputVolume(): %s", volume) );
			retorno.append(volume);	
		} catch (Exception e) {
			retorno.append(0f);
		}
        return retorno;
	}
	
	private MyBitSet alterarVolume(MyBitSet entrada) {
        MyBitSet volumeEntrada = entrada.slice(3, 32);
        System.out.println(volumeEntrada.toString() + ", valor: " + BinaryUtil.byteArrayToFloat(volumeEntrada.toByte(), false));
        
        float volume = BinaryUtil.byteArrayToFloat(volumeEntrada.toByte(), false);
        
        try {
        	Audio.setMasterOutputVolume(volume / 100.0f); // Linux - Audio
		} catch (Exception e) {
		}
        
        
        MyBitSet retorno = BinaryUtil.toMBitByte((byte)1, 1, false);
        
        try {
        	retorno.append(Audio.getMasterOutputVolume()); // float - 4 bytes = 32 bits	
		} catch (Exception e) {
			retorno.append(0f);
		}
        
        System.out.println(retorno.toString());	// 110111100010100010101110101000010
        
        return retorno;
	}
	
	private MyBitSet desligarPC() {
		String rt = "Desligar";
		MyBitSet retorno = BinaryUtil.toMBitByte((byte)0x1, 1, false);
		retorno.append(BinaryUtil.toMBit(rt));
		System.out.println("retorno: " + retorno.toString());
		
		Shutdown.shutDown();
		
		return retorno;
	}
	
	private MyBitSet mouseMove(MyBitSet entrada) {
		MyBitSet cod = entrada.slice(0, 3);
		System.out.println("cod: " + cod.toString() + " = " + BinaryUtil.toInt(cod, true)); // cod: 110 = 3
		
		MyBitSet wc = entrada.slice(3, 13);
		MyBitSet hc = entrada.slice(13+3, 13);
		MyBitSet xc = entrada.slice(13 + 13 + 3, 13);
		MyBitSet yc = entrada.slice(13 + 13 + 13 + 3, 13);
		System.out.println("wc: " + wc.toString() + " = " + BinaryUtil.toInt(wc, true));
		System.out.println("hc: " + hc.toString() + " = " + BinaryUtil.toInt(hc, true));
		System.out.println("xc: " + xc.toString() + " = " + BinaryUtil.toInt(xc, true));
		System.out.println("yc: " + wc.toString() + " = " + BinaryUtil.toInt(yc, true));
		
		int wcel = BinaryUtil.toInt(wc, true);
		int hcel = BinaryUtil.toInt(hc, true);
		int xcel = BinaryUtil.toInt(xc, true) + 50; // ajuste, android faz um -50 ?
		int ycel = BinaryUtil.toInt(yc, true) + 130; // ajuste, android faz um -130 ?

		Point screenSize = MouseControl.getScreenSize();
		int wpc = screenSize.x;
		int hpc = screenSize.y;
		
		int xPc = conversorXY(xcel, wcel, wpc);
		int yPc = conversorXY(ycel, hcel, hpc);

		//System.out.println("wpc: " + wpc + ", hpc: " + hpc + ", xPc: " + xPc + ", yPc: " + yPc);
		
		mc.moveMouse(xPc, yPc);
		
		// --- retorno
		String rt = "Recebido";
		MyBitSet retorno = BinaryUtil.toMBitByte((byte)0x1, 1, false);
		retorno.append(BinaryUtil.toMBit(rt));
		System.out.println("retorno: " + retorno.toString());
		return retorno;
	}
	
	private int conversorXY(int xyCel, int whCel, int whPc) {
		return (xyCel * whPc)/whCel;
	}
	
	private MyBitSet clickMouse() {
		String rt = "Click Recebido";
		MyBitSet retorno = BinaryUtil.toMBitByte((byte)0x1, 1, false);
		retorno.append(BinaryUtil.toMBit(rt));
		
		mc.clickMouse();
		
		System.out.println("retorno: " + retorno.toString());
		return retorno;
	}
	
	private MyBitSet lockScreen() {
		String rt = "Tela Bloqueada";
		MyBitSet retorno = BinaryUtil.toMBitByte((byte)0x1, 1, false);
		retorno.append(BinaryUtil.toMBit(rt));
		System.out.println("retorno: " + retorno.toString());
		
		LockUnLockScreen.lockSreen();

		return retorno;
	}

	private MyBitSet unLockScreen() {
		String rt = "Tela DesBloqueada";
		MyBitSet retorno = BinaryUtil.toMBitByte((byte)0x1, 1, false);
		retorno.append(BinaryUtil.toMBit(rt));
		System.out.println("retorno: " + retorno.toString());
		
		LockUnLockScreen.unLockSreen();

		return retorno;
	}
	
}

File: csharp/tcpserver_csharp/auxiliar/tratarrequests/TratarRequisicoesBin.cs
using System.Collections;
using System.Diagnostics;
using AudioSwitcher.AudioApi.CoreAudio;
using auxiliar.binarybits;
using tcpserver_csharp.auxiliar.desligarwindows;

namespace auxiliar.tratarrequests
{
    public class TratarRequisicoesBin
    {

        readonly BitArray codErro = BinaryBitsAux.to1Bit(false);
        readonly BitArray codOk = BinaryBitsAux.to1Bit(true);

        private readonly int mouseIncrement = 30;

        public BitArray tratar(BitArray entrada)
        {
            if (entrada == null || entrada.Count <= 0) { return msgErro(); }

            BitArray cod = BinaryBitsAux.splitBitArray(entrada, 0, 4);
            Console.WriteLine("cod: {0} = {1}", BinaryBitsAux.ToBitString(cod), BinaryBitsAux.toInt(cod, true));

            switch(BinaryBitsAux.toInt(cod, true)){
                case 0: return sinchronizar();
                case 1: return alterarVolume(entrada);
                case 2: return desligarPC();
                case 3: return mouseMove(entrada);
                case 4: return clickMouse();
                case 5: return lockScreen();
                case 6: return upMouse();
                case 7: return downMouse();
                case 8: return leftMouse();
                case 9: return rightMouse();
            }

            return msgErro();
        }

        private BitArray msgErro() {
            BitArray retorno = BinaryBitsAux.Combine(codErro, BinaryBitsAux.toBitArray("Erro"));
            Console.WriteLine("retorno: {0}", BinaryBitsAux.ToBitString(retorno));
            return retorno;
        }

        // 0 - sincronizar
        private BitArray sinchronizar()
        {
            CoreAudioDevice defaultPlaybackDevice = new CoreAudioController().DefaultPlaybackDevice;
            float volume = (float)(defaultPlaybackDevice.Volume);
            BitArray retorno = BinaryBitsAux.Combine(codOk, volume); // float - 4 bytes = 32 bits
            Console.WriteLine("retorno: {0}, volume: {1}, defaultPlaybackDevice.Volume: {2}", BinaryBitsAux.ToBitString(retorno), volume, defaultPlaybackDevice.Volume);
            return retorno;
        }

        // 1 - alterar o volume
        private BitArray alterarVolume(BitArray entrada)
        {
            BitArray volumeEntrada = BinaryBitsAux.splitBitArray(entrada, 4, 32);
            float volume = BinaryBitsAux.toFloat(volumeEntrada);
            Console.WriteLine("volumeEntrada: {0}, volume: {1}", BinaryBitsAux.ToBitString(volumeEntrada), volume);
            
            CoreAudioDevice defaultPlaybackDevice = new CoreAudioController().DefaultPlaybackDevice;
            defaultPlaybackDevice.Volume = volume;

            volume = (float)(defaultPlaybackDevice.Volume);
            BitArray retorno = BinaryBitsAux.Combine(codOk, volume); // float - 4 bytes = 32 bits
            Console.WriteLine("retorno: {0}, volume: {1}, defaultPlaybackDevice.Volume: {2}", BinaryBitsAux.ToBitString(retorno), volume, defaultPlaybackDevice.Volume);

            return retorno;
        }

        // 2 - desligar
        private BitArray desligarPC()
        {
            Console.WriteLine("Shutdown windows");

            DesligarWindows.Desligar();

            string rt = "Desligando";
            BitArray retorno = BinaryBitsAux.Combine(codOk, BinaryBitsAux.toBitArray(rt));
            Console.WriteLine("retorno: {0}", BinaryBitsAux.ToBitString(retorno));

            return retorno;
        }

        //3 - mouse
        private BitArray mouseMove(BitArray entrada)
        {
            int wc = BinaryBitsAux.toInt(BinaryBitsAux.splitBitArray(entrada, 4, 13), true);
            int hc = BinaryBitsAux.toInt(BinaryBitsAux.splitBitArray(entrada, 13 + 3, 13), true);
            int xc = BinaryBitsAux.toInt(BinaryBitsAux.splitBitArray(entrada, 13 + 13 + 3, 13), true) + 50; // ajuste, android faz um -50 ?
            int yc = BinaryBitsAux.toInt(BinaryBitsAux.splitBitArray(entrada, 13 + 13 + 13 + 3, 13), true)  + 50; // ajuste, android faz um -50 ?
            Console.WriteLine("wc: {0}, hc: {1}, xc: {2}, yc: {3}", wc, hc, xc, yc);

            if (wc < 0 || hc < 0 || xc < 0 || yc < 0) { return msgErro(); }

            int wpc = ScreenSize.getWidth();
            int hpc = ScreenSize.getHeight();
            wpc = wpc <= 0 ? 1920 : wpc;
            hpc = hpc <= 0 ? 1080 : (hpc + 50);

            int xPc = conversorXY(xc, wc, wpc);
            int yPc = conversorXY(yc, hc, hpc);
            MouseOperations.SetCursorPosition(xPc, yPc);

            string rt = "Recebido";
            BitArray retorno = BinaryBitsAux.Combine(codOk, BinaryBitsAux.toBitArray(rt));
            Console.WriteLine("retorno: {0}", BinaryBitsAux.ToBitString(retorno));

            return retorno;
        }

        private int conversorXY(int xyCel, int whCel, int whPc)
        {
            return (xyCel * whPc) / whCel;
        }

        // 4 - click mouse
        private BitArray clickMouse()
        {
            MouseOperations.MouseEvent(MouseOperations.MouseEventFlags.LeftDown);
            MouseOperations.MouseEvent(MouseOperations.MouseEventFlags.LeftUp);

            string rt = "Click Recebido";
            BitArray retorno = BinaryBitsAux.Combine(codOk, BinaryBitsAux.toBitArray(rt));
            Console.WriteLine("retorno: {0}", BinaryBitsAux.ToBitString(retorno));

            return retorno;
        }

        // 5 - Lock Screen
        private BitArray lockScreen(){
            LockScreen.LockWorkStation();

            string rt = "Tela Bloqueada";
            BitArray retorno = BinaryBitsAux.Combine(codOk, BinaryBitsAux.toBitArray(rt));
            Console.WriteLine("retorno: {0}", BinaryBitsAux.ToBitString(retorno));

            return retorno;
        }

        // 6 - Up Mouse
        private BitArray upMouse() {
            MouseOperations.MousePoint positionMouse = MouseOperations.GetCursorPosition();

            int posY = positionMouse.Y;
            posY -= mouseIncrement; if (posY < 0) { posY = 0; }
            MouseOperations.SetCursorPosition(positionMouse.X, posY);

            string rt = "Mouse Up";
            BitArray retorno = BinaryBitsAux.Combine(codOk, BinaryBitsAux.toBitArray(rt));
            Console.WriteLine("retorno: {0}", BinaryBitsAux.ToBitString(retorno));

            return retorno;
        }

        // 7 - Down Mouse
        private BitArray downMouse()
        {
            MouseOperations.MousePoint positionMouse = MouseOperations.GetCursorPosition();

            int posY = positionMouse.Y;
            posY += mouseIncrement; //if (posY >= hpc) { posY = hpc; }
            MouseOperations.SetCursorPosition(positionMouse.X, posY);

            string rt = "Mouse Down";
            BitArray retorno = BinaryBitsAux.Combine(codOk, BinaryBitsAux.toBitArray(rt));
            Console.WriteLine("retorno: {0}", BinaryBitsAux.ToBitString(retorno));

            return retorno;
        }

        // 8 - Left Mouse
        private BitArray leftMouse()
        {
            MouseOperations.MousePoint positionMouse = MouseOperations.GetCursorPosition();

            int posX = positionMouse.X;
            posX -= mouseIncrement; if (posX < 0) { posX = 0; }
            MouseOperations.SetCursorPosition(posX, positionMouse.Y);

            string rt = "Mouse Left";
            BitArray retorno = BinaryBitsAux.Combine(codOk, BinaryBitsAux.toBitArray(rt));
            Console.WriteLine("retorno: {0}", BinaryBitsAux.ToBitString(retorno));

            return retorno;
        }

        // 9 - Right Mouse
        private BitArray rightMouse()
        {
            MouseOperations.MousePoint positionMouse = MouseOperations.GetCursorPosition();

            int posX = positionMouse.X;
            posX += mouseIncrement; //if (posX >= wpc) { posX = wpc; }
            MouseOperations.SetCursorPosition(posX, positionMouse.Y);

            string rt = "Mouse Right";
            BitArray retorno = BinaryBitsAux.Combine(codOk, BinaryBitsAux.toBitArray(rt));
            Console.WriteLine("retorno: {0}", BinaryBitsAux.ToBitString(retorno));

            return retorno;
        }

    }

}